### DISCUSSION TITLE
Missing `From<&str>` implementation when using custom span struct

### DISCUSSION THREAD

**Discussion by mainrs** [zesterer/chumsky#225] [Q&A] [answered]
I have been using a custom span implementation and have issues when working with it. For some reason my code does not compile anymore. Somewhere a `From<&str>` implementation is missing and after skimming the code I do not really understand where this error comes from.

The error is triggered in [this](https://github.com/mainrs/s-expr-rs/blob/961dfa64dd2f7a6693f0c576a900650b466e1fc0/crates/parser/src/parser.rs#L94) line.

This is the error message:

```console
$ cargo test -- --nocapture
error[E0277]: the trait bound `Stream<'_, char, span::Span, _>: From<&str>` is not satisfied
   --> crates/parser/src/parser.rs:94:34
    |
94  |         let res = symbol().parse(i);
    |                            ----- ^ the trait `From<&str>` is not implemented for `Stream<'_, char, span::Span, _>`
    |                            |
    |                            required by a bound introduced by this call
    |
    = help: the following other types implement trait `From<T>`:
              <Stream<'a, T, std::ops::Range<usize>, Box<(dyn Iterator<Item = (T, std::ops::Range<usize>)> + 'a)>> as From<&'a [T; N]>>
              <Stream<'a, T, std::ops::Range<usize>, Box<(dyn Iterator<Item = (T, std::ops::Range<usize>)> + 'a)>> as From<&'a [T]>>
              <Stream<'a, T, std::ops::Range<usize>, Box<(dyn Iterator<Item = (T, std::ops::Range<usize>)> + 'a)>> as From<Vec<T>>>
              <Stream<'a, T, std::ops::Range<usize>, Box<(dyn Iterator<Item = (T, std::ops::Range<usize>)> + 'a)>> as From<[T; N]>>
              <Stream<'a, char, std::ops::Range<usize>, Box<(dyn Iterator<Item = (char, std::ops::Range<usize>)> + 'a)>> as From<&'a str>>
              <Stream<'a, char, std::ops::Range<usize>, Box<(dyn Iterator<Item = (char, std::ops::Range<usize>)> + 'static)>> as From<String>>
    = note: required because of the requirements on the impl of `Into<Stream<'_, char, span::Span, _>>` for `&str`
note: required by a bound in `parse`
   --> /home/me/.cargo/registry/src/github.com-1ecc6299db9ec823/chumsky-0.8.0/src/lib.rs:236:12
    |
236 |         S: Into<Stream<'a, I, <Self::Error as Error<I>>::Span, Iter>>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `parse`
```

Any help is appreciated.

**Comment by zesterer** [comment, marked as answer]
The reason that this doesn't work is that chumsky doesn't know how to assign instances of your custom span type to characters in the string. As you can see in the impl [here](https://docs.rs/chumsky/latest/chumsky/stream/struct.Stream.html#impl-From%3C%26%27a%20str%3E), `From<&str> for Stream` is only defined for the `Range<usize>` span.

To solve this, you'll need to use [`Stream::from_iter`](https://docs.rs/chumsky/latest/chumsky/stream/struct.Stream.html#method.from_iter). It's a relatively simple process, and the [existing implementations](https://docs.rs/chumsky/latest/src/chumsky/stream.rs.html#251-327) in the crate should give you an idea of how to use it.

i.e: your final code should look like `symbol().parse(Stream::from_iter(...))`.

**Reply by mainrs** [reply]
Thank you! For those finding the discussion, here is my final code:

```rust
pub fn parse<'a, I>(s: &str) -> (Option<Vec<Spanned<ast::AstNode>>>, Vec<chumsky::error::Simple<char, span::Span>>) {
    let len = s.len();
    let iter = Stream::from_iter(
        Span::new(len, len),
        Box::new(s.chars().enumerate().map(|(i,c)| (c, Span::new(i, i+1)))),
    );
    parser::parser().parse_recovery(iter)
}
```
